### PR TITLE
perf(Dgraph): Write xidmap to badger at the end.

### PR DESCRIPTION
**Pull request by martinmr**
Fixes DGRAPH-2514.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6665)

 
 
 

Docs Preview: [Dgraph Preview](https://dgraph-9410575f2c-103557.surge.sh)

### CODE REVIEW

File: xidmap/xidmap.go [hunk 1]
```diff
@@ -251,9 +237,29 @@ func (m *XidMap) AllocateUid() uint64 {
...
+	glog.Infof("Writing xid map to DB")
+	defer func() {
+		glog.Infof("Finished writing xid map to DB")
+	}()
+
 	for _, shard := range m.shards {
+		var err error
+		if m.writer != nil {
+			shard.Lock()
+			err = shard.trie.Iterate(func(key string, uid uint64) error {
```
Comment (NamanJain8) [inline xidmap/xidmap.go:249]: We are not doing error checking? `err` may be overwritten to `nil` by further entries?

**Review by NamanJain8 — COMMENTED** [review]
Looks good to me.

File: xidmap/trie.go [hunk 2]
```diff
@@ -128,3 +134,33 @@ func (t *Trie) put(offset uint32, key string, uid uint64) uint32 {
...
+	if err := t.iterate(n.left, prefix, fn); err != nil {
+		return err
+	}
+
+	if n.uid != 0 {
+		if err := fn(prefix+string(n.r), n.uid); err != nil {
+			return err
+		}
+	}
+	if err := t.iterate(n.mid, prefix+string(n.r), fn); err != nil {
```
Comment (jarifibrahim) [inline xidmap/trie.go:157]: hey @NamanJain8 , why do we iterate the n.mid with `prefix+string(n.r)` while the left and right are iterated with `prefix`?

File: xidmap/xidmap.go (hunk 1 above)
Comment (NamanJain8) [reply xidmap/xidmap.go:249]: My bad.

**Review by jarifibrahim — APPROVED** [review]

**Pull request merged**